### PR TITLE
Provide a way to change the state of a CircuitBreaker

### LINKED ISSUE #4010: Provide a way to change the state of a CircuitBreaker
So that users can dynamically close/open the circuit breaker to initiate particular downstream behavior as @batmanIsIronMan suggested.
Also, it's useful when we are testing.
Reference: https://github.com/Netflix/Hystrix/wiki/Configuration#circuitBreaker.forceOpen

### PR DESCRIPTION
**Pull request by jrhee17**
Motivation:

Users may want to manipulate `CircuitBreaker`'s internal state for testing or dynamically manipulating behavior.
Equivalents of these methods exist in `failsafe`, `resilience4j`, and `Hystrix`

Modifications:

- add a public method `CircuitBreaker#enterState`
- Add `CircuitState#FORCED_OPEN` which should be functionally equivalent to `CircuitState#OPEN`, but doesn't transition unless the user explicitly calls `CircuitBreaker#enterState`

Result:

- Closes #4010
- Users have greater control over circuitbreaker state

cc. @batmanIsIronMan @koeber99

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreaker.java [hunk 1]
```diff
@@ -73,4 +73,24 @@ static CircuitBreaker ofDefaultName() {
      * Decides whether a request should be sent or failed depending on the current circuit state.
      */
     boolean canRequest();
+
+    /**
+     * Transitions to {@link CircuitState#OPEN}.
+     */
+    void transitionToOpen();
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreaker.java:80]: How about just `enterState(CircuitState)`?

Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreaker.java]: Good point! Done 👍

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreaker.java [hunk 2]
```diff
@@ -73,4 +73,10 @@ static CircuitBreaker ofDefaultName() {
      * Decides whether a request should be sent or failed depending on the current circuit state.
      */
     boolean canRequest();
+
+    /**
+     * Transitions to the specified {@link CircuitState}.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreaker.java:78]: 'Transition' is a noun. How about 'Enters the specifies ...'?

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreaker.java [hunk 3]
```diff
@@ -73,4 +73,10 @@ static CircuitBreaker ofDefaultName() {
...
+     * Note that this method will always re-initialize the internal state.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreaker.java:79]: Could you rephrase this so that it is more useful to the users? What actually happens to our users' experience?

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreaker.java (hunk 2 above)
Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreaker.java]: Updated the documentation to match the method name 😅

**Review by ikhoon — APPROVED** [review]
Thanks, @jrhee17!

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreaker.java [hunk 4]
```diff
@@ -73,4 +73,13 @@ static CircuitBreaker ofDefaultName() {
      * Decides whether a request should be sent or failed depending on the current circuit state.
      */
     boolean canRequest();
+
+    /**
+     * Enters the specified {@link CircuitState}. Note that even if the {@link CircuitBreaker}
+     * is already in the specified {@link CircuitState}, timeouts such as
+     * {@link CircuitBreakerConfig#circuitOpenWindow()} and {@link CircuitBreakerConfig#trialRequestInterval()}
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreaker.java:80]: `CircuitBreakerConfig` is not a public class unfortunately.

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/NonBlockingCircuitBreaker.java [hunk 5]
```diff
@@ -141,6 +146,16 @@ public boolean canRequest() {
         return true;
     }
 
+    @Override
+    public void enterState(CircuitState circuitState) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/NonBlockingCircuitBreaker.java:150]: nit: rnn?

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/NonBlockingCircuitBreaker.java [hunk 6]
```diff
@@ -141,6 +146,16 @@ public boolean canRequest() {
...
+        final State oldState = state.getAndUpdate(st -> newState(circuitState));
+        if (oldState.circuitState() == circuitState) {
+            return;
+        }
+        notifyStateChanged(circuitState);
+        return;
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/NonBlockingCircuitBreaker.java:156]: nit:
```suggestion
```

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/NonBlockingCircuitBreaker.java [hunk 7]
```diff
@@ -141,6 +146,16 @@ public boolean canRequest() {
         return true;
     }
 
+    @Override
+    public void enterState(CircuitState circuitState) {
+        final State oldState = state.getAndUpdate(st -> newState(circuitState));
+        if (oldState.circuitState() == circuitState) {
+            return;
+        }
+        notifyStateChanged(circuitState);
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/NonBlockingCircuitBreaker.java:155]: Shouldn't we also call `logStateTransition`?

**Review by minwoox — COMMENTED** [review]
It looks good. 😄 
Left some minor comments.

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/NonBlockingCircuitBreaker.java (hunk 5 above)
Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/client/circuitbreaker/NonBlockingCircuitBreaker.java:150]: didn't understand this comment 😅

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/circuitbreaker/NonBlockingCircuitBreaker.java:150]: It should be **r**equire**N**on**N**ull. 🤣

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreaker.java (hunk 4 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreaker.java:80]: @jrhee17, you marked this as resolved but it seems not? :thinking:

Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreaker.java]: Sorry 😅  First this happened to me, but seems like `git push` updated the remote ref but didn't actually push changes to the server for some reason. Sorry about the confusion 🙏 

Screen Shot 2022-01-24 at 3 09 50 PM

**Review by trustin — APPROVED** [review]
Thanks, @jrhee17 :bow:

**Review by minwoox — APPROVED** [review]
Thanks! 👍

**Comment by ikhoon** [comment]
@jrhee17 Would you mind resolving the conflicts?

**Comment by ikhoon** [comment]
Nice work! @jrhee17

**Comment by ikhoon** [comment]
By the way, could you update the PR description?

**Comment by jrhee17** [comment]
Done 🙇

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitState.java [hunk 8]
```diff
@@ -33,5 +33,11 @@ public enum CircuitState {
      * If it doesn't complete within a certain time, another trial request will be sent again.
      * All other requests fails immediately same as OPEN.
      */
-    HALF_OPEN
+    HALF_OPEN,
```
Comment (martin-dd) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitState.java:36]: Where is FORCE_CLOSED state?

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitState.java:36]: I was wondering why you needed `FORCE_CLOSED`.
There is a situation where a request should be sent to a target even if the target is unavailable?

Comment (martin-dd) [reply core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitState.java:36]: Well edgecase of a severely misconfigured CB which keeps opening. Perhaps better than "FORCE_CLOSED" would be "DISABLED" - instead of force closing it, pretty much disable the CB and make it pass through

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitState.java:36]: Make sense. Would you mind opening an issue for that?
